handler.c

```diff
@@ -1,562 +1,546 @@
+  if (rc != noErr) {
+    fprintf(stderr,
+            "failed to set %s as handler "
+            "for %s:// (error %d)\n",
+            cb, cu, (int)rc);
+  }
+
+  return ((int)rc);
+}
 
-    if ( UTTypeConformsTo( uti, kUTTypeItem ) ||
-                UTTypeConformsTo( uti, kUTTypeContent ) ||
-                UTTypeConformsTo( uti, kUTTypeMessage ) ||
-                UTTypeConformsTo( uti, kUTTypeContact ) ||
-                UTTypeConformsTo( uti, kUTTypeArchive )) {
-        return( 1 );
-    }
+int fsethandler(char *spath) {
+  FILE *f = NULL;
+  char line[MAXPATHLEN * 2];
+  char *handler, *type, *role;
+  char **lineav = NULL;
+  int linenum = 0, rc = 0;
+  int len, htype;
+
+  if (spath != NULL) {
+    if ((f = fopen(spath, "r")) == NULL) {
+      fprintf(stderr, "fopen %s: %s\n", spath, strerror(errno));
+      return (1);
+    }
+  } else {
+    f = stdin;
+  }
+
+  while (fgets(line, sizeof(line), f) != NULL) {
+    linenum++;
+
+    len = strlen(line);
+    if (line[len - 1] != '\n') {
+      fprintf(stderr, "line %d: line too long\n", linenum);
+      exit(2);
+    }
+    line[len - 1] = '\0';
+
+    /* skip blanks and comments */
+    if (*line == '\0' || *line == '#') {
+      continue;
+    }
+
+    htype = parseline(line, &lineav);
+    switch (htype) {
+      case DUTI_TYPE_UTI_HANDLER:
+        handler = lineav[0];
+        type = lineav[1];
+        role = lineav[2];
+        break;
+
+      case DUTI_TYPE_URL_HANDLER:
+        handler = lineav[0];
+        type = lineav[1];
+        break;
+
+      default:
+        fprintf(stderr, "line %d: parsing failed\n", linenum);
+        rc = 1;
+        continue;
+    }
+
+    if (htype == DUTI_TYPE_UTI_HANDLER) {
+      duti_handler_set(handler, type, role);
+    } else if (htype == DUTI_TYPE_URL_HANDLER) {
+      duti_handler_set(handler, type, NULL);
+    }
+  }
+  if (ferror(f)) {
+    perror("fgets");
+    rc = 1;
+  }
+  if (fclose(f) != 0) {
+    perror("fclose");
+    rc = 1;
+  }
+
+  return (rc);
+}
 
-    return( 0 );
+int psethandler(char *spath) {
+  CFDictionaryRef plist;
+  CFArrayRef dharray;
+  CFDictionaryRef dhentry;
+  CFStringRef bid, uti, role, url_scheme;
+  CFIndex count, index;
+
+  int rc = 0;
+  int htype = DUTI_TYPE_UTI_HANDLER;
+  char handler[MAXPATHLEN], type[MAXPATHLEN];
+  char crole[255];
+
+  if (!spath) {
+    fprintf(stderr, "%s: invalid argument supplied\n", __FUNCTION__);
+    return (1);
+  }
+
+  if (read_plist(spath, &plist) != 0) {
+    fprintf(stderr, "%s: failed to read plist\n", __FUNCTION__);
+    return (1);
+  }
+
+  if (!plist) {
+    fprintf(stderr, "%s: Invalid plist\n", __FUNCTION__);
+    return (1);
+  }
+
+  if ((dharray = CFDictionaryGetValue(plist, DUTI_KEY_SETTINGS)) == NULL) {
+    fprintf(stderr, "%s is missing the settings array\n", spath);
+    CFRelease(plist);
+    return (1);
+  }
+
+  count = CFArrayGetCount(dharray);
+  for (index = 0; index < count; index++) {
```

Comment:
No, C style loops initialize as part of the loop declaration. But I suspect you know that.